keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/IntermittentFailureTests.kt

```diff
@@ -0,0 +1,170 @@
+
+  class Fixture {
+    val resourceRepository = InMemoryResourceRepository()
+    val artifactRepository = mockk<ArtifactRepository>()
+    val deliveryConfigRepository = mockk<DeliveryConfigRepository>()
+    val diffFingerprintRepository = InMemoryDiffFingerprintRepository()
+    val actuationPauser: ActuationPauser = mockk() {
+      coEvery { isPaused(any<String>()) } returns false
+      coEvery { isPaused(any<Resource<*>>()) } returns false
+    }
+    val plugin1 = mockk<ResourceHandler<DummyResourceSpec, DummyResourceSpec>>(relaxUnitFun = true)
+    val plugin2 = mockk<ResourceHandler<DummyResourceSpec, DummyResourceSpec>>(relaxUnitFun = true)
+    val publisher = mockk<ApplicationEventPublisher>(relaxUnitFun = true)
+    val clock = MutableClock()
+    val vetoRepository = InMemoryUnhappyVetoRepository(clock)
+
+    val dynamicConfigService: DynamicConfigService = mockk(relaxUnitFun = true) {
+      every {
+        // mimicking how a cluster is set up
+        getConfig(Int::class.java, "veto.unhappy.max-diff-count", any())
+      } returns 2
+      every {
+        // mimicking how a cluster is set up
+        getConfig(String::class.java, "veto.unhappy.waiting-time", any())
+      } returns "PT0S" // Duration.ZERO
+    }
+
+    val veto = UnhappyVeto(
+      resourceRepository,
+      diffFingerprintRepository,
+      vetoRepository,
+      dynamicConfigService,
+      "PT10M"
+    )
+    val vetoEnforcer = VetoEnforcer(listOf(veto))
+    val subject = ResourceActuator(
+      resourceRepository,
+      artifactRepository,
+      deliveryConfigRepository,
+      diffFingerprintRepository,
+      listOf(plugin1, plugin2),
+      actuationPauser,
+      vetoEnforcer,
+      publisher,
+      Clock.systemUTC())
+    val desired = DummyResourceSpec(data = "fnord")
+    val current = DummyResourceSpec()
+    val diff = DefaultResourceDiff(desired, current)
+  }
+
+  fun tests() = rootContext<Fixture> {
+
+    fixture { Fixture() }
+
+    before {
+      every { plugin1.name } returns "plugin1"
+      every { plugin1.supportedKind } returns SupportedKind(parseKind("plugin1/foo@v1"), DummyResourceSpec::class.java)
+      every { plugin2.name } returns "plugin2"
+      every { plugin2.supportedKind } returns SupportedKind(parseKind("plugin2/bar@v1"), DummyResourceSpec::class.java)
+    }
+
+    after {
+      resourceRepository.dropAll()
+    }
+
+    context("resource has a diff") {
+      val resource: Resource<DummyResourceSpec> = resource(
+        kind = parseKind("plugin1/foo@v1")
+      )
+      val deliveryConfig = deliveryConfig(resource)
+
+      before {
+        resourceRepository.store(resource)
+        resourceRepository.appendHistory(ResourceCreated(resource))
+        resourceRepository.appendHistory(ResourceValid(resource))
+        coEvery { plugin1.actuationInProgress(resource) } returns false
+        // diff has happened once
+        diffFingerprintRepository.store(resource.id, diff)
```

Comment:
I feel like this is forcing part of the behavior you want to verify. Who calls this functions in a real flow? I think it would be best to call the higher-level component which triggers this call.

Reply:
It's just making it so I don't call `subject.checkResource(resource)` again - I don't think I need to test that again. Trying to only reproduce the behavior I'm testing, and not test a bunch of other things. I'd say this is part of the setup of the test

Reply:
I understand, but what happens if the behavior of `subject.checkResource(resource)` later changes and stops storing the diff for some reason (including a regression)? Then you won't catch those bugs because you're emulating the expected behavior here, not letting the actual implementation do the setup for you.

Reply:
but I'm not testing that `subject.checkResource(resource)` stores the diff